components/iosnoop/src/biosnoop.c

```diff
@@ -0,0 +1,62 @@
+#include <uapi/linux/ptrace.h>
+#include <linux/blkdev.h>
+
+struct data_t {
+    u32 pid;
+    u64 rwflag;
+    u64 len;
+};
+
+struct val_t {
+    u32 pid;
+};
+
+BPF_HASH(infobyreq, struct request *, struct val_t);
+BPF_PERF_OUTPUT(events);
+
+int trace_pid_start(struct pt_regs *ctx, struct request *req)
+{
+    struct val_t val = {};
+    u64 ts;
+
+    u64 id = bpf_get_current_pid_tgid();
+    u32 pid = id & (((u64)1 << 32) - 1);
+    u32 tgid = id >> 32;
+    if (tgid != ##TGID##) {
+        return 0;
+    }
+    val.pid = pid;
+    infobyreq.update(&req, &val);
+    return 0;
+}
+
+// output
+int trace_req_completion(struct pt_regs *ctx, struct request *req)
+{
+    struct data_t data = {};
+    struct val_t *valp;
+
+    valp = infobyreq.lookup(&req);
+    if (valp == 0) return 0;
+    data.pid = valp->pid;
+    data.len = req->__data_len;
+
+/*
+ * The following deals with a kernel version change (in mainline 4.7, although
+ * it may be backported to earlier kernels) with how block request write flags
+ * are tested. We handle both pre- and post-change versions here. Please avoid
+ * kernel version tests like this as much as possible: they inflate the code,
+ * test, and maintenance burden.
+ */
+#ifdef REQ_WRITE
+    data.rwflag = !!(req->cmd_flags & REQ_WRITE);
+#elif defined(REQ_OP_SHIFT)
+    data.rwflag = !!((req->cmd_flags >> REQ_OP_SHIFT) == REQ_OP_WRITE);
+#else
+    data.rwflag = !!((req->cmd_flags & REQ_OP_MASK) == REQ_OP_WRITE);
+#endif
+
+    events.perf_submit(ctx, &data, sizeof(data));
```

Comment:
Definitely a kernel-to-user copy here. I'm not sure the actual overhead when read frequency is normally below 10K/s, but I'd prefer we periodically read out a trace summary every 10 ms or so. I've seen use case that accesses `BPF_HASH` from outside the bpf, maybe we can do it here as well?

Reply:
Let's see how large the overhead is. And if reading out the summary periodically, it can hardly get an accurate byte for one thread like this:
```
let ctx = IOContext::new();
do_some_read();
let read_bytes = ctx.delta();
```

Reply:
I haven't read the code thoroughly, but I believe the update to global map might lag behind the actual IO operation (it must be async as to not affect user bio? not sure), which means the `ctx.delta()` can be less useful here. Or if by any chance it's synchronously updated, the overhead of acquiring one global mutex can't be easy to swallow.

Reply:
Yes, the global map is updated asynchronously. I'm trying to make it synchronously.

Reply:
I rethink about it and maybe the precise tracing for one request isn't necessary? To throttle a read, all we need to know is *approximately* how many bytes has been read during some period.

Reply:
I agree we don't need to trace per query stat. Instead we can pass the IOType into bcc code, and store a BPF map of (IOType, read|write) => bytes. We could collect the map every 10ms maybe and use it for the rate limiter.

We can also have a BPF_HISTOGRAM per IO type to export query latency.

Reply:
okay, it would be great not to trace per query. 
For the latest changes(not push yet), it collects the bytes in bpf hashmap and there is no bfp poll in tikv thread more. Instead, it gets the bytes stats from bpf hashmap(syscall) everytime calling IOContext. So rate limiter just need to call `IOContext::delta()` for any period.

Reply:
Have no idea about how to passing `IOType` into bcc code.
I think we can let every thread to collect read|write from IOContext periodically and the threads know the IOType of their own. Then the threads push the stats to rate limiter?

Reply:
@Connor1996 I found a `bpf_probe_read_user` function, not sure if it's helpful.

Reply:
Can the IO type be pass through a BPF_HASH map?

Reply:
No, BPF_HASH only supports kernel to user copy, with bpf_probe_read_user maybe we can pass a pointer of thread_local IOType into bpf.

Reply:
BPF_HASH supports user to kernel. We can pass the tls of pid to that BPF_HASH and check io type in bcc.
See https://github.com/iovisor/bcc/blob/0b2f4d0b981803ea8d8393ad2eec93c699e59df1/examples/cpp/pyperf/PyPerfUtil.cc#L212

Reply:
That's great! I was reading this issue https://github.com/iovisor/bcc/issues/2015, if it does work out maybe we can close it.

Reply:
seems it is supported at initial. https://github.com/iovisor/bcc/blob/a1a1775bef3cc9739566110093b3d85d82fa2f95/src/cc/libbpf.h#L42
I'm not sure if I'm thinking wrong. But rust-bcc does support `table.set()` too. https://docs.rs/bcc/0.0.30/bcc/table/struct.Table.html#impl